ign-rviz-common/src/rviz/common/frame_manager.cpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2020 Sarathkrishnan Ramesh
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ignition/rviz/common/frame_manager.hpp"
+
+#include <string>
+#include <utility>
+#include <memory>
+#include <vector>
+
+namespace ignition
+{
+namespace rviz
+{
+namespace common
+{
+FrameManager::FrameManager(rclcpp::Node::SharedPtr node)
+{
+  this->node = std::move(node);
+
+  tfBuffer = std::make_shared<tf2_ros::Buffer>(this->node->get_clock());
+  tfListener = std::make_shared<tf2_ros::TransformListener>(*tfBuffer);
+
+  this->subscriber = this->node->create_subscription<tf2_msgs::msg::TFMessage>(
+    "/tf", 10,
+    std::bind(&FrameManager::tf_callback, this, std::placeholders::_1));
+}
+
+void FrameManager::setFixedFrame(std::string fixedFrame)
+{
+  this->fixedFrame = fixedFrame;
+}
+
+std::string FrameManager::getFixedFrame()
+{
+  return this->fixedFrame;
+}
+
+void FrameManager::getFrames(std::vector<std::string> & frames)
+{
+  tfBuffer->_getFrameStrings(frames);
+}
+
+void FrameManager::tf_callback(tf2_msgs::msg::TFMessage::SharedPtr msg)
+{
+  std::lock_guard<std::mutex>(this->tf_mutex_);
+
+  if (this->fixedFrame.empty()) {
+    RCLCPP_ERROR(this->node->get_logger(), "No frame id specified");
+    return;
+  }
+
+  std::vector<std::string> frame_ids;
+  tfBuffer->_getFrameStrings(frame_ids);
+
+  builtin_interfaces::msg::Time timeStamp = msg->transforms[0].header.stamp;
+  timePoint =
+    tf2::TimePoint(
+    std::chrono::seconds(timeStamp.sec) +
+    std::chrono::nanoseconds(timeStamp.nanosec));
+
+  for (const auto frame : frame_ids) {
+    try {
+      geometry_msgs::msg::TransformStamped tf = tfBuffer->lookupTransform(
+        fixedFrame, timePoint,
+        frame, timePoint,
+        fixedFrame, tf2::Duration(5000));
```

Comment:
time is too long

Reply:
The visualization isn't that smooth if I reduce this number. Can you help me fix this?

Reply:
add a comment just explaining why this number is big. and maybe a TODO(Sarath18) to revisit this

Reply:
Got it